### PR TITLE
Add `ConditionalKeys`, `ConditionalPick` and `ConditionalExcept` types

### PR DESCRIPTION
**Pull request by ifiokjr**
### Description

Add types to conditionally pick and exclude properties from any shape based on the provided `Condition`. 

This type comes in handy quite a lot in my library code bases where transforming types or picking properties based on conditions helps dynamically drive automatic inference. 

Any suggestions for better names is welcome. Since this code base provides the `Except` type I've used the name `ConditionalExcept`. However, it might be more understandable to use `ConditionalOmit`.

### CODE REVIEW

File: source/conditionals.d.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  d: {}
+}
+
+type StringsKeysOnly = ConditionalKeys<Example, string>;
+// => 'a'
+```
+
+To support partial types make sure your `Condition` is a union of undefined as shown below.
+*/
+export type ConditionalKeys<Base, Condition> = NonNullable<
```
Comment (resynth1943) [inline source/conditionals.d.ts]:
```suggestion
export type ConditionalKeys<Base, Condition> = NonNullable<{
```

**Comment by resynth1943** [comment]
> Any suggestions for better names is welcome. Since this code base provides the `Except` type I've used the name `ConditionalExcept`. However, it might be more understandable to use `ConditionalOmit`.

Nah. `ConditionalExcept` should be fine here, as it relies on the `Except` type, not the `Omit` type.

File: source/conditionals.d.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+Pick the keys from any typescript shape which match the provided `Condition`.
+
+This is useful when you want to create a new type from a specific subset from an already created type. For example you might want to pick all the primitive properties from a class and form a new automatically derived type.
+
+@example
+```
+import {Primitive, ConditionalPick} from 'type-fest';
+
+class Awesome {
+  string name;
```
[the remark this answers is not in the document]
Comment (ifiokjr) [reply source/conditionals.d.ts]: Thinking in [dart](https://dart.dev/).

**Comment by ifiokjr** [comment]
I'll address the requested changes once I have a few moments.

File: source/conditionals.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (resynth1943) [reply source/conditionals.d.ts]: I was starting to learn that one day. It looks really good 😛

File: test-d/conditionals.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+expectType<{a: string; c?: string}>(exampleConditionalPickWithUndefined);
+
+declare const exampleConditionalExcept: ConditionalExcept<Example, string>;
+expectType<{b?: string | number; c?: string; d: Record<string, unknown>}>(exampleConditionalExcept);
+
+declare const awesomeConditionalExcept: ConditionalExcept<Awesome, Primitive>;
+expectType<{ run: () => void }>(awesomeConditionalExcept);
+
+declare const exampleConditionalExceptWithUndefined: ConditionalExcept<Example, string | undefined>;
+expectType<{b?: string | number; d: Record<string, unknown> }>(exampleConditionalExceptWithUndefined);
```
[the remark this answers is not in the document]
Comment (ifiokjr) [reply test-d/conditionals.ts]: `xo` fails if I add the trailing `;` for one line shapes. I've removed this but kept the suggested space padding.

File: source/conditionals.d.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
+import {Except} from './except';
```
[the remark this answers is not in the document]
Comment (ifiokjr) [reply source/conditionals.d.ts]: I've split it up into seperate files.

**Comment by ifiokjr** [comment]
I've addressed the requested changes with some minor deviations as explained above.

**Comment by resynth1943** [comment]
Merged in 1... 2... 3... 😛

**Comment by sindresorhus** [comment]
Thanks, @ifiokjr  👌

**Pull request merged**